torch/nn/utils/parametrize.py

```diff
@@ -53,72 +55,211 @@ def cached():
+        here for completeness. It shall not be instantiated by the user.
 
     Args:
-        modules (iterable): an iterable of modules representing the parametrizations
+        modules (sequence): sequence of modules representing the parametrizations
         original (Parameter or Tensor): parameter or buffer that is parametrized
     """
     original: Tensor
 
     def __init__(
-        self, modules: Iterable[Module], original: Union[Tensor, Parameter]
+        self, modules: Sequence[Module], original: Union[Tensor, Parameter]
     ) -> None:
+        # We require this because we need to treat differently the first parametrization
+        # This should never throw, unless this class is used from the outside
+        if len(modules) == 0:
+            raise ValueError("ParametrizationList requires one or more modules.")
+
         super().__init__(modules)
-        if isinstance(original, Parameter):
-            self.register_parameter("original", original)
-        else:
-            self.register_buffer("original", original)
 
-    def set_original_(self, value: Tensor) -> None:
-        r"""This method is called when assigning to a parametrized tensor.
+        # In plain words:
+        # module.weight must keep its dtype and shape.
+        # Furthermore, if there is no right_inverse or the right_inverse returns a tensor,
+        # this should be of the same dtype as the original tensor
+        #
+        # We check that the following invariants hold:
+        #    X = module.weight
+        #    Y = param.right_inverse(X)
+        #    assert isinstance(Y, Tensor) or
+        #           (isinstance(Y, collections.Sequence) and all(isinstance(t, Tensor) for t in Y))
+        #    Z = param(Y) if isisntance(Y, Tensor) else param(*Y)
+        #    # Consistency checks
+        #    assert X.dtype == Z.dtype and X.shape == Z.shape
+        #    # If it has one input, this allows to be able to use set_ to be able to
+        #    # move data to/from the original tensor without changing its id (which is what the
+        #    # optimiser uses to track parameters)
+        #    if isinstance(Y, Tensor)
+        #      assert X.dtype == Y.dtype
+        # Below we use original = X, new = Y
+
+        original_shape = original.shape
+        original_dtype = original.dtype
+
+        # Compute new
+        with torch.no_grad():
+            new = original
+            for module in reversed(self):  # type: ignore[call-overload]
+                if hasattr(module, "right_inverse"):
+                    new = module.right_inverse(new)
+                # else, we assume that right_inverse is the identity
+
+        if not isinstance(new, Tensor) and not isinstance(new, collections.Sequence):
+            raise ValueError("'right_inverse' must return a Tensor or a Sequence of tensors (list, tuple...). "
+                             f"Got {type(new).__name__}")
+
+        # If it is a sequence of one tensor, we unpack it
```

Comment:
Do we really want to do that? Also this is not documented ?

Reply:
The idea was that in the 1 input case we can reuse the original tensor, which is good. Now, this is such a weird case... What do you think we should do?

Reply:
If people return (foo,) instead of foo, I'm sure they have a good reason to do it. So I would lean towards removing this.
Also the way it works right now, you will pass "foo" to their forward and not "(foo,)" which is not consistent.